File: contrib/A093211.py
divm = 11
n = 11
last = 10**n
ssts = []
debug = 0

def setgoodlist(len, adiv):
    atmp = [0,1,2,3,4,5,6,7,8,9]
    btmp = [0,1,2,3,4,5,6,7,8,9]
    for i in atmp:
        remain = (adiv - i*(10 **(len-1))) % adiv
        if (adiv - 10*remain) % adiv > 9:
            btmp.remove(i)
    if debug > 0:
        print("btmp=" + str(btmp));
    return btmp

def dropdigs(k,l):
    ktmp = (k // l) * l
    ktmp -= 1
    while (ktmp % divm) != 0:
        ktmp -= 1
    if debug >= 2:
        print("dropdigs(" + str(k) + "," + str(l) + " -> " + str(ktmp))
    return ktmp

def walking(k):
    aa = 0
    kt = k*10
    ktl = kt % last
    if debug >= 2:
        print("walk in(" + str(k) + "), last=" + str(last) + ", ktl=" + str(ktl))
    if ((ktl - 1) % divm) + 10 < divm:
        if debug >= 2:
            print("walk premature(" + str(k) + "), aa=" + str(aa))
        return aa
    if ktl % divm == 0:
        a = kt
    else:
        a = kt + (divm - (ktl % divm))

    al = a % last
    if al not in ssts:
        if debug >= 2:
            print("walk !contains(" + str(al) + "), a=" + str(a))
        ssts.append(al)
        aa = max(aa,a)
        atmp = walking(a)
        aa = max(aa,atmp)
        ssts.remove(al)
    else:
        if debug >= 2:
            print("walk contains(" + str(al) + "), a=" + str(a))
    if debug >= 2:
        print("walk out(" + str(k) + ") -> " + str(aa))
    return aa
    

for n in range(1,16):
    goodlist = setgoodlist(n,divm)
    last = 10**n
    beg = int(n*'9')
    end = int((n-1)*"9")
    i = beg
    while i % divm != 0:
        i -= 1
    if debug >= 2:
        print ("i=" + str(i) + ", n=" + str(n) + ", beg=" + str(beg) + ", end=" + str(end))
    an = i
    oldan = an
    anlen = n
    while i > end:
        ssts = [i]
        if i % 100000 == 0:
            anlen = len(str(an))
            if anlen > 2*n:
                anlen = 2*n - 1
        an = max(an,walking(i))
        i -= divm
        for j in range(anlen - n+1):
            jten = 10 ** (n - j-1)
            if jten < 1:
                break
            while (i // jten) % 10 not in goodlist:
                i = dropdigs(i,jten)
    print(str(n) + "   " + str(an))

File: contrib/A093211.pl
use strict;
use integer;

my $divm = shift(@ARGV);
my $n = $divm;
my $last = 10**$n;
my %ssts = ();
my $debug = 0;

sub setgoodlist() {
  my ($len, $adiv) = @_;
  my @btmp = (1,1,1,1,1 ,1,1,1,1,1); # 0..9
  foreach my $i (0..9) {
    my $remain = ($adiv - $i * (10**($len - 1))) % $adiv;
    if (($adiv - 10 * $remain) % $adiv > 9) {
      $btmp[$i] = 0; # remove
    }
  } # for $i
  if ($debug >= 2) { print join(", ", @btmp) . "\n"; }
  return @btmp;
} # sub setgoodlist

sub dropdigs() {
  my ($k, $l) = @_;
  my $ktmp = ($k / $l) * $l;
  $ktmp --;
  while (($ktmp % $divm) != 0) {
    $ktmp --;
  }
  if ($debug >= 2) { print "dropdigs($k,$l) -> $ktmp\n"; }
  return $ktmp;
} # dropdigs

sub walking() {
  my ($k) = @_;
  my $aa = 0;
  my $kt = $k * 10;
  my $ktl = $kt % $last;
  if ((($ktl - 1) % $divm) + 10 < $divm) {
    return $aa;
  }
  my $a;
  if ($ktl % $divm == 0) {
    $a = $kt;
  } else {
    $a = $kt + ($divm - ($ktl % $divm));
  }

  my $al = $a % $last;
  if (! defined($ssts{$al})) {
    $ssts{$al} = 1;
    if ($a > $aa) {
      $aa = $a;
    }
    my $atmp = &walking($a);
    if ($atmp > $aa) {
      $aa = $atmp;
    }
    delete $ssts{$al};
  }
  if ($debug >= 4) {
    print "walking($k) -> $aa\n";
  }
  return $aa;
} # walking

for ($n = 2; $n < 12; $n ++) { # main loop
  my @goodlist = &setgoodlist($n, $divm);
  $last = 10**$n;
  my $beg = '9' x $n;
  my $end = '9' x ($n-1);
  if ($debug >= 2) { print "n=$n, beg=$beg, end=$end\n"; }
  my $i = $beg;
  while ($i % $divm != 0) {
    $i -= 1;
  }
  my $an = $i;
  my $oldan = $an;
  my $anlen = $n;
  while ($i > $end) {
    %ssts = ($i, 1);
    if ($i % 100000 == 0) {
      $anlen = length($an);
      if ($anlen > 2*$n) {
        $anlen = 2*$n - 1;
      }
    }
    my $wi = &walking($i);
    if ($wi > $an) {
      $an = $wi;
    }
    $i -= $divm;
    for (my $j = 0; $j < $anlen - $n + 1; $j ++) {
      my $jten = 10 ** ($n - $j - 1);
      if ($jten < 1) {
        last;
      }
      while ($goodlist[($i / $jten) % 10] == 0) {
        $i = &dropdigs($i, $jten);
      }
    } # for j
  } # while i
  print "$n $an\n";
} # main loop
